### PR TITLE
[WIP] Introduce proxy registration protocol version 2

### PR DESCRIPTION
**Pull request by ekohl**
Part of https://community.theforeman.org/t/exposing-capabilities-in-the-smart-proxy/9860

### CODE REVIEW

File: db/migrate/20180705230311_smart_proxy_capabilities.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    change_table :smart_proxy_features do |t|
+      t.primary_key :id
+      t.text :capabilities, array: true
+      if t.respond_to?(:jsonb)
+        t.jsonb :settings
+      elsif t.respond_to?(:json)
+        t.json :settings
+      else
+        t.text :settings
+      end
```
Comment (lzap) [inline db/migrate/20180705230311_smart_proxy_capabilities.rb:13]: Implementation of JSON support in ActiveRecord is poor, we shall avoid this as long as possible. There is no real benefit in storing the data in native JSON, you don't query any data and ActiveRecord is actually slower than just with text implementation.

In other words, unless there is a JSON index created and a JSON query in real use, there is no added benefit in storing JSON natively through ActiveRecord as of today. Note that the SQL JSON syntax is not standardized, that's also a challenge we'd need to solve.

Unless you show good reasons to do that, let's stick with text.

**Comment by lzap** [comment]
I like where this is going, what do you want to do with settings again?

File: db/migrate/20180705230311_smart_proxy_capabilities.rb (hunk 1 above)
Comment (ekohl) [reply db/migrate/20180705230311_smart_proxy_capabilities.rb:13]: Perhaps that was the result of wanting to play with new technology on a Friday. It was still on my list to evaluate this and perhaps fall back to a simple serialized JSON string like on sqlite for reasons you mentioned.

**Comment by ekohl** [comment]
@ehelms wants to use this to store the the Pulp API URL rather than assuming the Pulp hostname matches the Proxy. Another example that I think would improve it is replacing the live request for TFTP:

https://github.com/theforeman/foreman/blob/4dfba1a1645c22df405fdc648b16348080424bba/lib/proxy_api/tftp.rb#L41-L51

We need to be a bit more careful there but we can probably also remove the ugly ip -> reverse dns -> ip part there. That should make it all a bit more straight forward and predictable.

There might be more settings that Foreman could know about and use, otherwise it's mostly going to be empty :)

**Comment by jlsherrill** [comment]
New PR has been open here: https://github.com/theforeman/foreman/pull/6314

**Pull request closed without merging**